{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58860297",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import random\n",
    "import math\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "4f507c89",
   "metadata": {},
   "outputs": [],
   "source": [
    "class WrightFisher():\n",
    "    \"\"\"\n",
    "    Haploid Wright-Fisher model\n",
    "    \n",
    "    Parameters:\n",
    "    N: population size\n",
    "    s: selection coefficient\n",
    "    n: initial variant allele count\n",
    "    \"\"\"\n",
    "    def __init__(self, N, s = 0, n = 1):\n",
    "        # Set population size, selection coefficient and allele count for A\n",
    "        self.N = N\n",
    "        self.s = s\n",
    "        self.n = n\n",
    "        \n",
    "        # Set initial allele frequencies\n",
    "        self.p0 = self.n/self.N\n",
    "        self.q0 = self.N - self.p0\n",
    "        \n",
    "        # Set fitnesses relative to selection\n",
    "        self.wa = 1 + s\n",
    "        self.wA = 1\n",
    "        \n",
    "    # Run Wright-Fisher simulation for a single site\n",
    "    def simulateSingleSite(self, max_generations):\n",
    "        # Create array of population alleles per generation\n",
    "        population_alleles = np.zeros(max_generations + 1, dtype = int)\n",
    "        # Set allele counts for initial generation\n",
    "        population_alleles[0] = self.n\n",
    "        \n",
    "        # Initialise counters\n",
    "        gen = 0\n",
    "        allele_count = self.n\n",
    "        max_allele_counts = self.n\n",
    "        \n",
    "        # Repeat for n generations\n",
    "        while (gen < max_generations and 0 < allele_count < self.N + 1):\n",
    "            gen += 1\n",
    "            \n",
    "            # Get the allele count in the previous generation\n",
    "            allele_count = population_alleles[gen - 1]\n",
    "            # Calculate probability of sampling the allele \n",
    "            prob = allele_count * (1 + self.s) / (allele_count * (1 + self.s) + self.N - allele_count)\n",
    "            \n",
    "            # Simulate number of copies in next generation\n",
    "            population_alleles[gen] = np.random.binomial(n = self.N,\n",
    "                                                         p = prob)\n",
    "        \n",
    "        return population_alleles\n",
    "    \n",
    "    # Run Wright-Fisher simulation for two loci\n",
    "    def simulateTwoSites(self, alleles, allele_counts, singleton_haploid, r):\n",
    "        # Whether one loci's allele is fixed\n",
    "        allele_fixed = False\n",
    "\n",
    "        # Set initial population\n",
    "        current_alleles = alleles\n",
    "        new_allele_counts = allele_counts\n",
    "        \n",
    "        # Record the haplotype frequencies per generation\n",
    "        haplotype_frequencies = np.array(list(alleles.values()))\n",
    "\n",
    "        # Record D', D and r^2 per generation\n",
    "        ld_results = self.linkageDisequilibrium(alleles, allele_counts, singleton_haploid)\n",
    "        D_normalised_per_gen = ld_results[\"D'\"]\n",
    "        D_per_gen = ld_results[\"D\"]\n",
    "        r2_per_gen = ld_results[\"r2\"]\n",
    "        \n",
    "        # Get list of alleles, i.e. A, a, B, b\n",
    "        allele_names = np.array(list(alleles.keys()))\n",
    "        # Bias the probability of selection of each haplotype by the selection coefficient\n",
    "        selection_coefficient_prob = np.array([-self.s, -self.s, -self.s, -self.s])\n",
    "\n",
    "        # Get indexes of haplotypes to give add selection coefficient\n",
    "        positive_haplotypes = [a for a in allele_names if singleton_haploid[0] in a]\n",
    "        haplotype_indexes = np.concatenate((np.where(allele_names == positive_haplotypes[0]),\n",
    "                                            np.where(allele_names == positive_haplotypes[1]))).flatten()\n",
    "        # Set A allele from the singleton haploid to have positive selection \n",
    "        selection_coefficient_prob[haplotype_indexes] = self.s\n",
    "\n",
    "        # Run simulation until one loci's allele is fixed\n",
    "        while not allele_fixed:\n",
    "            # Set new empty population\n",
    "            new_alleles = {\"AB\": 0, \"Ab\": 0, \"aB\": 0, \"ab\": 0}\n",
    "\n",
    "            # Populate the new population\n",
    "            for i in range(self.N):\n",
    "\n",
    "                # Probability of each haplotype being selected\n",
    "                selection_probability = np.array(list(current_alleles.values())) / self.N\n",
    "                \n",
    "                # Add selection coefficient and normalise to add to 1\n",
    "                selection_probability = selection_probability + selection_coefficient_prob\n",
    "                selection_probability = (selection_probability - np.min(selection_probability)) / np.ptp(selection_probability)\n",
    "                selection_probability = selection_probability / selection_probability.sum()\n",
    "\n",
    "                # Randomly select a haplotype from the population\n",
    "                individual = np.random.choice(list(current_alleles.keys()),\n",
    "                                              p = selection_probability)\n",
    "\n",
    "                # No recombination\n",
    "                if np.random.uniform() > r:\n",
    "                    # Offspring is the same as the individual in the next generation\n",
    "                    new_alleles[individual] += 1\n",
    "\n",
    "                # Make a recombinant haplotype\n",
    "                else:\n",
    "                    # Randomly select an allele to keep\n",
    "                    allele_1 = np.random.choice(list(individual))\n",
    "\n",
    "                    # If A loci, randomly select a B allele from the population\n",
    "                    if allele_1.lower() == \"a\":\n",
    "                        allele_2 = np.random.choice([\"B\", \"b\"],\n",
    "                                                    p = np.array([new_allele_counts[\"B\"],\n",
    "                                                                  new_allele_counts[\"b\"]]) / self.N)\n",
    "                        # Form the new recombinant haplotype\n",
    "                        offspring = allele_1 + allele_2\n",
    "\n",
    "                    # Otherwise randomly select an A allele from the population\n",
    "                    else:\n",
    "                        allele_2 = np.random.choice([\"A\", \"a\"],\n",
    "                                                    p = np.array([new_allele_counts[\"A\"],\n",
    "                                                                  new_allele_counts[\"a\"]]) / self.N)\n",
    "                        offspring = allele_2 + allele_1\n",
    "\n",
    "                    # Add the recombinant offspring to the population\n",
    "                    new_alleles[offspring] += 1\n",
    "\n",
    "            # Update the new allele counts for the new population\n",
    "            new_allele_counts = {\"A\": new_alleles[\"AB\"] + new_alleles[\"Ab\"],\n",
    "                                 \"a\": new_alleles[\"aB\"] + new_alleles[\"ab\"],\n",
    "                                 \"B\": new_alleles[\"AB\"] + new_alleles[\"aB\"],\n",
    "                                 \"b\": new_alleles[\"Ab\"] + new_alleles[\"ab\"]}\n",
    "\n",
    "            # Record D' and r^2\n",
    "            ld_results = self.linkageDisequilibrium(new_alleles, new_allele_counts, singleton_haploid)\n",
    "            D_normalised_per_gen = np.append(D_normalised_per_gen, ld_results[\"D'\"])\n",
    "            D_per_gen = np.append(D_per_gen, ld_results[\"D\"])\n",
    "            r2_per_gen = np.append(r2_per_gen, ld_results[\"r2\"])\n",
    "            \n",
    "            # Record the haplotype frequency for the population\n",
    "            haplotype_frequencies = np.vstack((haplotype_frequencies,\n",
    "                                               np.array(list(new_alleles.values()))))\n",
    "\n",
    "            # Check if any of the alleles have fixed\n",
    "            if self.N in new_allele_counts.values():\n",
    "                # Stop loop as an allele at a loci was lost\n",
    "                allele_fixed = True\n",
    "            else:\n",
    "                # Set alleles for new population\n",
    "                current_alleles = new_alleles\n",
    "                \n",
    "        return {\"D'\": D_normalised_per_gen, \"D\": D_per_gen, \"r2\": r2_per_gen,\n",
    "                \"haplotype_frequencies\": np.transpose(haplotype_frequencies)}\n",
    "        \n",
    "    # Set population sizes for two loci simulation\n",
    "    def initialiseTwoLoci(self, distribution):\n",
    "        site_distribution = distribution[2:-2]\n",
    "        \n",
    "        # Population allele counts\n",
    "        alleles = {\"AB\": 0, \"Ab\": 0, \"aB\": 0, \"ab\": 0}\n",
    "\n",
    "        # Calculate probability of samples allele count between 2 to N - 2\n",
    "        prob = site_distribution / sum(site_distribution)\n",
    "\n",
    "        A_allele_count = np.random.choice(a = range(2, len(site_distribution) + 2), p = prob)\n",
    "        a_allele_count = self.N - A_allele_count\n",
    "\n",
    "        # Randomly introduce a singleton at a loci\n",
    "        singleton_A = np.random.choice([\"A\", \"a\"])\n",
    "        singleton_B = np.random.choice([\"B\", \"b\"])\n",
    "\n",
    "        if singleton_B == \"B\":\n",
    "            B_allele_count = 1\n",
    "            b_allele_count = self.N - 1\n",
    "        else:\n",
    "            B_allele_count = self.N - 1\n",
    "            b_allele_count = 1\n",
    "\n",
    "        # Record the singleton\n",
    "        alleles[singleton_A + singleton_B] = 1\n",
    "\n",
    "        # Set the remaining allele counts\n",
    "        if singleton_A + singleton_B == \"AB\":\n",
    "            alleles[\"Ab\"] = A_allele_count - 1\n",
    "            alleles[\"aB\"] = B_allele_count - alleles[\"AB\"]\n",
    "            alleles[\"ab\"] = self.N - sum(alleles.values())\n",
    "\n",
    "        elif singleton_A + singleton_B == \"Ab\":\n",
    "            alleles[\"AB\"] = A_allele_count - 1\n",
    "            alleles[\"ab\"] = b_allele_count - alleles[\"Ab\"]\n",
    "            alleles[\"aB\"] = self.N - sum(alleles.values())\n",
    "\n",
    "        elif singleton_A + singleton_B == \"aB\":\n",
    "            alleles[\"AB\"] = B_allele_count - 1\n",
    "            alleles[\"Ab\"] = A_allele_count - alleles[\"AB\"]\n",
    "            alleles[\"ab\"] = self.N - sum(alleles.values())\n",
    "\n",
    "        elif singleton_A + singleton_B == \"ab\":\n",
    "            alleles[\"AB\"] = A_allele_count\n",
    "            alleles[\"Ab\"] = b_allele_count - alleles[\"ab\"]\n",
    "            alleles[\"aB\"] = self.N - sum(alleles.values())\n",
    "            \n",
    "        # Return the population genomes and allele frequencies\n",
    "        return {\"alleles\": alleles,\n",
    "                \"allele_counts\": {\"A\": A_allele_count,\n",
    "                                  \"a\": a_allele_count,\n",
    "                                  \"B\": B_allele_count,\n",
    "                                  \"b\": b_allele_count},\n",
    "                \"singleton_haploid\": singleton_A + singleton_B}\n",
    "    \n",
    "    # Calculate D' and r^2 LD statistics\n",
    "    def linkageDisequilibrium (self, alleles, allele_counts, singleton_haploid):\n",
    "        # Calculate proportion of individuals with an allele at each locus, i.e. pA, pa, pB, pb\n",
    "        allele_probabilities = dict(zip([\"p\" + i for i in allele_counts.keys()],\n",
    "                                        [allele_counts[j] / 100 for j in allele_counts.keys()]))\n",
    "        # Proportion of two alleles occuring at the two loci, i.e. pAB, pAb, paB, pab\n",
    "        haploid_probabilities = dict(zip([\"p\" + i for i in alleles.keys()],\n",
    "                                         [alleles[j] / 100 for j in alleles.keys()]))\n",
    "\n",
    "        # Calculate coefficient of linkage disequilibrium for each haplotype\n",
    "        D_haplotype = {\"AB\": haploid_probabilities[\"pAB\"] - (allele_probabilities[\"pA\"] * allele_probabilities[\"pB\"]),\n",
    "                       \"Ab\": haploid_probabilities[\"pAb\"] - (allele_probabilities[\"pA\"] * allele_probabilities[\"pb\"]),\n",
    "                       \"aB\": haploid_probabilities[\"paB\"] - (allele_probabilities[\"pa\"] * allele_probabilities[\"pB\"]),\n",
    "                       \"ab\": haploid_probabilities[\"pab\"] - (allele_probabilities[\"pa\"] * allele_probabilities[\"pb\"])}\n",
    "\n",
    "        # Calculate normalised D by dividing by the maximum difference between observed and expected haplotype frequencies\n",
    "        try:\n",
    "            D_normalised = D_haplotype[singleton_haploid] / max(D_haplotype.values())\n",
    "        except:\n",
    "            # Prevent division by zero\n",
    "            D_normalised = None\n",
    "\n",
    "        # Calculate genetric correlation coefficient between pairs of loci\n",
    "        try:\n",
    "            r2 = (D_haplotype[singleton_haploid] ** 2) / (allele_probabilities[\"p\" + singleton_haploid[0]] *\n",
    "                                                          (1 - allele_probabilities[\"p\" + singleton_haploid[0]]) *\n",
    "                                                          allele_probabilities[\"p\" + singleton_haploid[1]] *\n",
    "                                                          (1 - allele_probabilities[\"p\" + singleton_haploid[1]]))\n",
    "        except:\n",
    "            r2 = None\n",
    "        \n",
    "        return {\"D'\": D_normalised, \"D\": D_haplotype[singleton_haploid], \"r2\": r2}\n",
    "    \n",
    "    # Plot distribution of maximum allele count\n",
    "    def plotMaxAllele(self, max_allele_counts,\n",
    "                      title = \"Distribution of Maximum Allele Number Reached ($\\mathregular{M_i}$)\"):\n",
    "        max_allele_fig = plt.figure(figsize = (10, 5))\n",
    "        max_allele_hist = plt.hist(max_allele_counts,\n",
    "                                   color = \"orange\",\n",
    "                                   edgecolor = \"black\", \n",
    "                                   linewidth = 1.2,\n",
    "                                   bins = min(max(max_allele_counts), 100))\n",
    "        plt.xlabel(\"Maximum Allele Number ($\\mathregular{M_i}$)\") \n",
    "        plt.ylabel(\"Number of Simulations\") \n",
    "        plt.title(title)\n",
    "        plt.show()\n",
    "        \n",
    "        return max_allele_fig\n",
    "        \n",
    "    # Plot distribution of time for simulations where the variant fixes or is lost\n",
    "    def plotTime(self, fixation_times, loss_times, combine_plots = False,\n",
    "                 title = \"Distribution of Time Steps ($\\mathregular{T_i}$) for the Allele to Fix or be Lost\"):\n",
    "        \n",
    "        # Check whether to plot data on the same axis\n",
    "        if combine_plots:\n",
    "            time_hist_fig = plt.figure(figsize = (10, 5))\n",
    "            # Plot fixation histogram\n",
    "            plt.hist(fixation_times, color = \"lime\", label = \"Fixed\",\n",
    "                     edgecolor = \"black\", linewidth = 1.2,\n",
    "                     bins = 50, alpha = 0.5)\n",
    "            # Also plot loss histogram on same axis\n",
    "            plt.hist(loss_times, color = \"red\", label = \"Lost\",\n",
    "                     edgecolor = \"black\", linewidth = 1.2,\n",
    "                     bins = 50, alpha = 0.5)\n",
    "            plt.xlabel(\"Time ($\\mathregular{T_i}$)\") \n",
    "            plt.ylabel(\"Number of Simulations\")\n",
    "            plt.title(title)\n",
    "            plt.legend(loc = \"upper right\")\n",
    "            \n",
    "        # Otherwise create subplots\n",
    "        else:\n",
    "            time_hist_fig, axes = plt.subplots(1, 2, figsize = (12, 5))\n",
    "            axes[0].hist(fixation_times, color = \"lime\", label = \"Fixed\",\n",
    "                         edgecolor = \"black\", linewidth = 1.2,\n",
    "                         bins = 50, alpha = 0.5)\n",
    "            axes[1].hist(loss_times, color = \"red\", label = \"Lost\",\n",
    "                         edgecolor = \"black\", linewidth = 1.2,\n",
    "                         bins = 50, alpha = 0.5)\n",
    "            axes[0].title.set_text(\"Fixation\")\n",
    "            axes[1].title.set_text(\"Loss\")\n",
    "            time_hist_fig.suptitle(title, y = 1.03)\n",
    "            \n",
    "            for ax in axes:\n",
    "                ax.set_xlabel(\"Time ($\\mathregular{T_i}$)\")\n",
    "                ax.set_ylabel(\"Number of Simulations\")\n",
    "            \n",
    "        plt.show()\n",
    "        \n",
    "        return time_hist_fig\n",
    "    \n",
    "    # Plot density heatmap of allele counts over time\n",
    "    def plotPathDensity(self, path_density, title = \"Path Density\", \n",
    "                        fontsize = 12, cmap = \"viridis\",\n",
    "                        logscale = True, add_expected = False,\n",
    "                        expected_colour = \"white\"):\n",
    "        \n",
    "        # Create figure\n",
    "        density_fig, ax = plt.subplots(figsize = (11, 9))\n",
    "        # Log the data if logscale set as True\n",
    "        density_data = np.log(path_density + 1) if logscale else path_density\n",
    "        \n",
    "        # Add density heatmap of allele counts over time\n",
    "        heatmap = ax.imshow(density_data, cmap = cmap, aspect = \"auto\")\n",
    "        \n",
    "        # Flip the y-axis so n increases upwards\n",
    "        ax.set_ylim(ax.get_ylim()[::-1])\n",
    "        \n",
    "        # Add colour bar\n",
    "        colour_bar = density_fig.colorbar(heatmap, pad = 0.15 if add_expected else 0.05)\n",
    "        colour_bar_label = \"Log(Density + 1)\" if logscale else \"Density\"\n",
    "        colour_bar.ax.set_title(colour_bar_label, fontsize = 10)\n",
    "\n",
    "        if add_expected:\n",
    "            # Create sigmoid expectation curve for p(t) under the deterministic, no drift model\n",
    "            t = np.linspace(0, len(simulation_results[\"path_density\"][0]) - 1,\n",
    "                            num = len(simulation_results[\"path_density\"][0]))\n",
    "            n = 1/(1 + ((self.q0/self.p0) * ((self.wA / self.wa) ** t)))\n",
    "            \n",
    "            # Plot curve on top\n",
    "            ax2 = plt.twinx()\n",
    "            sns.lineplot(x = t, y = n, linewidth = 3, ax = ax2, color = expected_colour)\n",
    "            ax.axis('tight')\n",
    "            ax2.set_ylabel(\"Expected p(t) with no Drift\", fontsize = fontsize)\n",
    "        \n",
    "        ax.set_xlabel(\"Time (t)\", fontsize = fontsize)\n",
    "        ax.set_ylabel(\"Allele Count (n)\", fontsize = fontsize)\n",
    "        ax.set_title(title, fontsize = fontsize + 2)\n",
    "        plt.show()\n",
    "        \n",
    "        return density_fig\n",
    "        \n",
    "    # Plot distribution of number of copies of an allele over all generations\n",
    "    def plotAlleleCount(self, all_allele_counts, logscale = True,\n",
    "                        title = \"Distribution of Total Allele Copies Across Generations S(n)\"):\n",
    "        allele_counts_fig = plt.figure(figsize = (10, 5))\n",
    "        allele_counts_bar = plt.bar(range(len(all_allele_counts)),\n",
    "                                    all_allele_counts,\n",
    "                                    color = \"skyblue\",\n",
    "                                    edgecolor = \"black\", \n",
    "                                    linewidth = 1.2)\n",
    "        plt.xlabel(\"Allele Count (n)\") \n",
    "        plt.ylabel(\"Frequency\")\n",
    "        plt.title(title)\n",
    "        plt.yscale(\"log\" if logscale else \"linear\")\n",
    "        plt.show()\n",
    "        \n",
    "        return allele_counts_fig\n",
    "    \n",
    "    # Run the single site simulator n times\n",
    "    def repeatSimulateSingle(self, max_generations = 500, n_times = 1000):\n",
    "        # The highest numbers the allele reaches per simulation\n",
    "        max_allele_counts = np.zeros(n_times, dtype = int)\n",
    "        # The times taken for the allele to fix or be lost\n",
    "        end_times = np.zeros(n_times, dtype = int)\n",
    "        # Density of allele counts over time for n simulations\n",
    "        path_density = np.zeros((self.N + 1, max_generations), dtype = int)\n",
    "        \n",
    "        # Total time for simulations that fix and loose the allele\n",
    "        fixation_times = np.array([])\n",
    "        loss_times = np.array([])\n",
    "        fluctuating_times = np.array([])\n",
    "        # Average time for simulations that fix and loose the allele\n",
    "        mean_fixation_time = None\n",
    "        mean_loss_time = None\n",
    "        \n",
    "        # Run the simulation n times\n",
    "        for i in range(n_times):\n",
    "            # Run Wright-Fisher model\n",
    "            population_alleles = self.simulateSingleSite(max_generations)\n",
    "            # Get the maximum allele count and simulation end time\n",
    "            max_allele_counts[i] = max(population_alleles)\n",
    "            end_times[i] = len(population_alleles[population_alleles != 0])\n",
    "            \n",
    "            # Increment path density for the new allele counts over time\n",
    "            for j in range(1, len(population_alleles)):\n",
    "                path_density[population_alleles[j]][j - 1] += 1\n",
    "                \n",
    "            # Check if allele fixes\n",
    "            if population_alleles[-1] == self.N:\n",
    "                # Add time when allele fixes\n",
    "                fixation_times = np.append(fixation_times,\n",
    "                                           np.where(population_alleles == self.N)[0][0])\n",
    "            elif population_alleles[-1] == 0:\n",
    "                # Add time taken for allele to be lost\n",
    "                loss_times = np.append(loss_times, end_times[i])\n",
    "                \n",
    "        # Number of simulations in which the allele fixes, is lost or still fluctuating\n",
    "        n_fixed = len(fixation_times)\n",
    "        n_lost = len(loss_times)\n",
    "        n_fluctuating = n_times - n_fixed - n_lost\n",
    "        \n",
    "        # Calculate average fixation and loss time\n",
    "        if n_fixed > 0:\n",
    "            mean_fixation_time = sum(fixation_times) / len(fixation_times)\n",
    "        if n_lost > 0:\n",
    "            mean_loss_time = sum(loss_times) / len(loss_times)\n",
    "            \n",
    "        # S(n) distribution, i.e. number of copies of the variant alelle over all generations\n",
    "        all_allele_counts = np.array([sum(row) for row in path_density])\n",
    "        \n",
    "        return {\"max_allele_counts\": max_allele_counts,\n",
    "                \"end_times\": end_times,\n",
    "                \"path_density\": path_density,\n",
    "                \"n_fixed\": n_fixed,\n",
    "                \"n_lost\": n_lost,\n",
    "                \"n_fluctuating\": n_fluctuating,\n",
    "                \"mean_fixation_time\": mean_fixation_time,\n",
    "                \"mean_loss_time\": mean_loss_time,\n",
    "                \"fixation_times\": fixation_times,\n",
    "                \"loss_times\": loss_times,\n",
    "                \"all_allele_counts\": all_allele_counts}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9eb1d9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set selection coefficient s\n",
    "selection_coefficient = 0.005\n",
    "\n",
    "# Run the simulator\n",
    "wf_simulator = WrightFisher(N = 100, s = selection_coefficient)\n",
    "simulation_results = wf_simulator.repeatSimulateSingle(n_times = 1000)\n",
    "\n",
    "# Print summary statistics\n",
    "print(\"Fixed \" + str(simulation_results[\"n_fixed\"]) + \" times\")\n",
    "print(\"Lost \" + str(simulation_results[\"n_lost\"]) + \" times\")\n",
    "print(\"Fluctuating \" + str(simulation_results[\"n_fluctuating\"]) + \" times\")\n",
    "if simulation_results[\"mean_fixation_time\"] != None:\n",
    "    print(\"Average fixation time of \" + str(round(simulation_results[\"mean_fixation_time\"], 3)))\n",
    "print(\"Average loss time of \" + str(round(simulation_results[\"mean_loss_time\"], 3)))\n",
    "\n",
    "# Add s to plot titles and file names if s not zero\n",
    "extra_title = \" for s = \" + str(selection_coefficient) if selection_coefficient != 0 else \"\"\n",
    "extra_file_name = \"_s=\" + str(selection_coefficient) if selection_coefficient != 0 else \"\"\n",
    "\n",
    "# Plot results\n",
    "allele_count_bar = wf_simulator.plotAlleleCount(simulation_results[\"all_allele_counts\"], logscale = True,\n",
    "                                               title = \"Distribution of Total Allele Copies Across Generations S(n)\" + extra_title)\n",
    "time_bars = wf_simulator.plotTime(simulation_results[\"fixation_times\"], simulation_results[\"loss_times\"],\n",
    "                                 title = \"Distribution of Time Steps ($\\mathregular{T_i}$) for the Allele to Fix or be Lost\" + extra_title)\n",
    "max_allele_hist = wf_simulator.plotMaxAllele(simulation_results[\"max_allele_counts\"],\n",
    "                                            title = \"Distribution of Maximum Allele Number Reached ($\\mathregular{M_i}$)\" + extra_title)\n",
    "density_plot = wf_simulator.plotPathDensity(simulation_results[\"path_density\"], \n",
    "                                            cmap = \"Spectral\", logscale = True,\n",
    "                                            title = \"Path Density for 1000 Simulations\" + extra_title,\n",
    "                                            add_expected = True if selection_coefficient != 0 else False)\n",
    "\n",
    "# Save plots\n",
    "allele_count_bar.savefig(\"Allele_Count_Bar\" + extra_file_name + \".pdf\", bbox_inches = \"tight\")\n",
    "time_bars.savefig(\"Time_Bars\" + extra_file_name + \".pdf\", bbox_inches = \"tight\")\n",
    "max_allele_hist.savefig(\"Max_Allele_Histogram\" + extra_file_name + \".pdf\", bbox_inches = \"tight\")\n",
    "density_plot.savefig(\"Path_Density\" + extra_file_name + \".pdf\", bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f09f120e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set new selection coefficient and recombination rate\n",
    "selection_coefficient = 0.005\n",
    "recombination_rate = 0.05\n",
    "\n",
    "# Create a new simulator\n",
    "wf_simulator2 = WrightFisher(N = 100, s = selection_coefficient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bc2a99e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial alleles: {'AB': 1, 'Ab': 15, 'aB': 0, 'ab': 84}\n",
      "Initial allele count: {'A': 16, 'a': 84, 'B': 1, 'b': 99}\n",
      "Singleton allele: AB\n"
     ]
    }
   ],
   "source": [
    "# Set number of times to test initial distribution of D' and r^2\n",
    "n_initial = 1\n",
    "\n",
    "# Record D' and r^2 for n initial populations\n",
    "initial_ld_d_norm = np.zeros(n_initial)\n",
    "initial_ld_r2 = np.zeros(n_initial)\n",
    "\n",
    "for i in range(n_initial):\n",
    "    # Set initial haplotype sizes, number of AB, Ab, aB, ab\n",
    "    simulation_results2 = wf_simulator.repeatSimulateSingle(n_times = 1000)\n",
    "    initial_loci = wf_simulator2.initialiseTwoLoci(simulation_results2[\"all_allele_counts\"])\n",
    "\n",
    "    # Get allele counts and which haplotype has a count of 1\n",
    "    initial_alleles = initial_loci[\"alleles\"]\n",
    "    initial_allele_counts = initial_loci[\"allele_counts\"]\n",
    "    singleton = initial_loci[\"singleton_haploid\"]\n",
    "\n",
    "    # Record LD measures\n",
    "    ld_results = wf_simulator2.linkageDisequilibrium(initial_alleles, initial_allele_counts, singleton)\n",
    "    initial_ld_d_norm[i] = ld_results[\"D'\"]\n",
    "    initial_ld_r2[i] = ld_results[\"r2\"]\n",
    "    \n",
    "# Set lists of haplotypes and alleles\n",
    "all_haplotypes = list(initial_alleles.keys())\n",
    "all_alleles = list(initial_allele_counts.keys())\n",
    "    \n",
    "print(\"Initial alleles: \" + str(initial_alleles))\n",
    "print(\"Initial allele count: \" + str(initial_allele_counts))\n",
    "print(\"Singleton allele: \" + str(singleton))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9dfa57bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAmcAAAFUCAYAAABhtJuPAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAdNElEQVR4nO3de5RlZX3m8e9DN3gDRAXlLjgDKo6AWqBZmgheElABNSaKBpVxhnjBkGiiJDGaDFlLnQlZyghhUBkxMeIlRFsD0YkKzqgYGkW0QbBFhRaE9gbYGkjDb/44p+VQnDq1q6ld9Xaf72etWnX23u/e51f1Vlc9/e7Lm6pCkiRJbdhmuQuQJEnSXQxnkiRJDTGcSZIkNcRwJkmS1BDDmSRJUkMMZ5IkSQ0xnElaVEl+JcmXklyU5INJtl3umiRpS2I4k7TYvgc8raqeClwDHLPM9UjSFsVwJm2lkqxJctjmbh9p990kz+j6vlV1fVX9Yri4EbhzwrEfmeSrSW5N8ntd32NL1/V7v5nHXlB/zdq3t7okdWc4k7YgC/nDW1WPqaoL59pvdHsfkuwLHAl8ckKzNwAXVtUOVXXaIrzniUlWJ7ktyfvGbH9wkn9MsiHJ95K8uMu2LttntX1Kki8muTnJj5N8Ickhm7b3/b3vYjl+JiR1s3K5C5C0dUmyAngAcA5wXFXdPqH5w4FzN+M9VlbVxjGbrgf+EvgN4H5jtp8O3A48DDgY+KckX6uqNfNsm2/f0dp2ZBBIXwV8GNgO+FXgtoV+nZKmkyNn0hZqOPLxh0kuH47QfCjJfWdtf0aSvwX2Bj6R5GdJ3jC6ffj65CTfHp5evCLJ8xZQx39J8ukk703yE+B1wAeBP6+qqybs91ngcOBdw7r2T/LoJBcm+enwFNvRs76eNya5HNiQ5B7/uayq86rqY8CPxrzfA4DfBP6sqn5WVf8PWAUcN2nbfPuO+dL2H9bywaq6o6p+UVWfrqrLZ30tzxh5/UfDftww/D4+LMkFw/74lyQPGtm3kvzHkeX3JfnLOb7HY/u148/EnH0x0nbsz9+wn74/fN+rkjx9XH2SxjOcSVu23waOAPYFDgRePrtBVR0HXAscVVXbV9V/H3OcbzMY3Xkg8BfA3yXZrWMNBwK/AnwceAhwE/BE4M3DP+4vHLdTVT0N+L/AiVW1PfAd4BPAp4GHAq8FPpDkkSO7HQs8G9hpjpGzSfYH7qiqq0fWfQ14zDzb5tt3tquBO5Kck+TI0WA1wW8Czxy+z1HABcCfADsz+D29udfjje3X+X4mMrjDdr6+gDE/f8M2JwKHVNUODEYxv7uZ9UtTyXAmbdlOG16A/2MGf0wP3pyDVNVHhse5s6o+BHwLOLTj7gcBf1VVq4b7n1NVO1fVYcOPD3U8zpOA7YG3VdXtVfVZBqcHjx1pc1pVXTdyw8FCbA/cPGvdzcAO82ybb9+7qapbgKcABbwbWJ9kVZKHTajtf1bVjVX1fQaB9ctV9dWqug34R+Bx831x49yLfu3SFzD+5+8O4D7AAUm2rarvVtW3N6d+aVoZzqQt2w9GXv+cwR/UBUvy0iSXDU9h/RT4TwxGbbo4EPjI5rzvLLsD11XV6N2d3wP2GFm+7l4c/2fAjrPW7QjcOs+2+fa9h6q6sqpeXlV7Mvhe7g68Y0JtN468/sWY5aXu1y59AWN+/qpqLfD7wJ8DNyU5N8num1O/NK0MZ9J0qLk2JHk4gxGeE4GHVNVOwDeAzHfQ4b7bAt9chBqvB/ZKMvp7aW/g+yPLc34dHVwNrEyy38i6g4A182ybb9+JquqbwPsYBKPF8HPg/iPLu45r1KFfJ30vu/TFnKrq76vqKQxu+Cjg7V32kzRgOJOmw43AI+bY9gAGf0DXAyQ5nu5B4iDg67NGWDbXl4ENwBuSbJvB87aOYgF3cyZZObwofQWwIsl9N904UFUbgPOA/5bkAUmezOABuX87adt8+46p4VFJXp9kz+HyXgxOB1688G/JWJcBL06yIskRwFPnaDdfv076mdjsvsjg2XVPS3If4N8YjPzdMd9+ku5iOJOmw1uBNw1Pb/3h6IaqugI4FfgSgz/YjwW+0PG4BzEIC/fa8JEbRzN4NtoPgTOAlw5Hnrp6E4MwcDLwO8PXbxrZ/moGj9i4icEdpa8aeRTGpG1dtm9yK4MbIr6cZAODUPYN4PUL+DomOYlBUPop8BLgY+MadejXST8T96Yv7gO8bbjfDxjcUPAn3b40SQCpujdnCSRJkrSYHDmTJElqiOFMkiSpIYYzSZKkhhjOJEmSGmI4kyRJaojhTJIkqSErl7uAhdp5551rn332We4yJEmS5nXppZf+sKp2Wcg+W1w422effVi9evVylyFJkjSvJN9b6D6e1pQkSWqI4UySJKkhhjNJkqSGGM4kSZIaYjiTJElqiOFMkiSpIYYzSZKkhhjOJEmSGtJbOEtydpKbknxjju1JclqStUkuT/L4vmqRJEnaUvQ5cvY+4IgJ248E9ht+nAD8TY+1SJIkbRF6C2dV9XngxxOaHAO8vwYuBnZKsltf9UiSJG0JlvOasz2A60aW1w3X3UOSE5KsTrJ6/fr1S1KcpK3XrrvuQ5K7fey66z7LXZYkAcsbzjJmXY1rWFVnVdVMVc3sssuCJnaXpHu48cbvMfh1c9fHYJ0kLb/lDGfrgL1GlvcErl+mWiRJkpqwnOFsFfDS4V2bTwJurqoblrEeSZKkZbeyrwMn+SBwGLBzknXAW4BtAarqTOB84FnAWuDnwPF91SJJkrSl6C2cVdWx82wv4DV9vb8kSdKWyBkCJEmSGmI4kyRJaojhTJIkqSGGM0mSpIYYziRJkhpiOJMkSWqI4UySJKkhhjNJkqSGGM4kSZIaYjiTJElqiOFMkiSpIYYzSZKkhhjOJEmSGmI4kyRJaojhTJIkqSGGM0mSpIYYziRJkhpiOJMkSWqI4UySJKkhhjNJkqSGGM4kSZIaYjiTJElqiOFMkiSpIYYzSZKkhhjOJEmSGmI4kyRJaojhTJIkqSGGM0mSpIYYziRJkhpiOJMkSWqI4UySJKkhhjNJkqSGGM4kSZIaYjiTJElqiOFMkiSpIYYzSZKkhhjOJEmSGmI4kyRJaojhTJIkqSGGM0mSpIYYziRJkhpiOJMkSWqI4UySJKkhvYazJEckuSrJ2iQnj9n+wCSfSPK1JGuSHN9nPZIkSa3rLZwlWQGcDhwJHAAcm+SAWc1eA1xRVQcBhwGnJtmur5okSZJa1+fI2aHA2qq6pqpuB84FjpnVpoAdkgTYHvgxsLHHmiRJkprWZzjbA7huZHndcN2odwGPBq4Hvg6cVFV3zj5QkhOSrE6yev369X3VK0mStOz6DGcZs65mLf8GcBmwO3Aw8K4kO95jp6qzqmqmqmZ22WWXxa5TkiSpGX2Gs3XAXiPLezIYIRt1PHBeDawFvgM8qseaJEmSmtZnOLsE2C/JvsOL/F8ErJrV5lrg6QBJHgY8Erimx5okSZKatrKvA1fVxiQnAp8CVgBnV9WaJK8cbj8TOAV4X5KvMzgN+saq+mFfNUmSJLWut3AGUFXnA+fPWnfmyOvrgV/vswZJkqQtiTMESJIkNcRwJkmS1BDDmSRJUkMMZ5IkSQ0xnEmSJDXEcCZJktQQw5kkSVJDDGeSJEkNMZxJkiQ1xHAmSZLUkHnDWZKTkuyYgfcm+UoSp1ySJEnqQZeRs/9cVbcwmANzF+B44G29ViVJkjSluoSzDD8/C/jfVfW1kXWSJElaRF3C2aVJPs0gnH0qyQ7Anf2WJUmSNJ1WdmjzCuBg4Jqq+nmShzA4tSlJkqRFNm84q6o7k9wIHJCkS5iTJEnSZpo3bCV5O/BC4ArgjuHqAj7fY12SJElTqctI2HOBR1bVbT3XIkmSNPW63BBwDbBt34VIkiSp28jZz4HLknwG+OXoWVX9Xm9VSZIkTaku4WzV8EOSJEk963K35jlJtgP2H666qqr+vd+yJEmSplOXuzUPA84BvstgZoC9krysqrxbU5IkaZF1Oa15KvDrVXUVQJL9gQ8CT+izMEmSpGnU5W7NbTcFM4Cquhrv3pQkSepFl5Gz1UneC/ztcPklwKX9lSRJkjS9uoSzVwGvAX6PwTVnnwfO6LMoSZKkadXlbs3bgL8efkiSJKlHc4azJB+uqt9O8nUGc2neTVUd2GtlkiRJU2jSyNlJw8/PWYpCJEmSNOFuzaq6Yfjy1VX1vdEP4NVLU54kSdJ06fIojWeOWXfkYhciSZKkydecvYrBCNkjklw+smkH4At9FyZJkjSNJl1z9vfABcBbgZNH1t9aVT/utSpJkqQpNWc4q6qbgZuBYwGSPBS4L7B9ku2r6tqlKVGSJGl6zHvNWZKjknwL+A5wEYMJ0C/ouS5JkqSp1OWGgL8EngRcXVX7Ak/Ha84kSZJ60SWc/XtV/QjYJsk2VfU54OB+y5IkSZpOXebW/GmS7RnMqfmBJDcBG/stS5IkaTp1GTk7BvgF8AfAPwPfBo7qsyhJkqRp1WXi8w0ji+f0WIskSdLUm/QQ2lsZM+E5EKCqasfeqpIkSZpSk+bW3KGqdhzzsUPXYJbkiCRXJVmb5OQ52hyW5LIka5JctLlfiCRJ0tZg3tOaSfYet36+h9AmWQGczmBuznXAJUlWVdUVI212As4Ajqiqa4cPupUkSZpaXe7W/KeR1/cF9gWuAh4zz36HAmur6hqAJOcyuLngipE2LwbO2xT0quqmjnVLkiRtlbrcEPDY0eUkjwd+t8Ox9wCuG1leBzxxVpv9gW2TXMhgQvV3VtX7OxxbkiRpq9Rl5OxuquorSQ7p0DTjdh/z/k9gMOvA/YAvJbm4qq6+24GSE4ATAPbee+xZVkmSpK1Cl2vOXjeyuA3weGB9h2OvA/YaWd4TuH5Mmx8OH9exIcnngYOAu4WzqjoLOAtgZmZm3B2kkiRJW4UuD6HdYeTjPgyuQTumw36XAPsl2TfJdsCLgFWz2nwc+NUkK5Pcn8Fpzyu7Fi9JkrS16XLN2V9szoGramOSE4FPASuAs6tqTZJXDrefWVVXJvln4HLgTuA9VfWNzXk/SZKkrUGqJp8lTDID/CnwcEbCXFUd2G9p483MzNTq1auX460lbSWScM9LYMN8vw8laaGSXFpVMwvZp8sNAR8A/gj4OoPRLUmSJPWkSzhbX1WzrxWTJElSD7qEs7ckeQ/wGeC2TSur6rzeqpIkSZpSXcLZ8cCjgG2567RmAYYzSZKkRdYlnB00e5YASZIk9aPLc84uTnJA75VIkiSp08jZU4CXJfkOg2vOAtRyPUpDkiRpa9YlnB3RexWSJEkCJoSzJDtW1S3ArUtYjyRJ0lSbNHL298BzgEsZ3J2ZkW0FPKLHuiRJkqbSnOGsqp4z/Lzv0pUjSZI03ea8WzPJw5M8cGT58CTvTPIHSbZbmvIkSZKmy6RHaXwYeABAkoOBjwDXAgcDZ/RdmCRJ0jSadM3Z/arq+uHr3wHOrqpTk2wDXNZ7ZZIkSVNo0sjZ6A0AT2MwtyZVdef45pIkSbq3Jo2cfTbJh4EbgAcBnwVIshtw+xLUJkmSNHUmhbPfB14I7AY8par+fbh+V+BPe65LkiRpKk16lEYB545Z/9VeK5IkSZpiXSY+lyRJ0hIxnEmSJDVk0kNoPzP8/PalK0eSJGm6TbohYLckTwWOTnIud3+0BlX1lV4rkyRJmkKTwtmbgZOBPYG/nrWtGDz7TJIkSYto0t2aHwU+muTPquqUJaxJkiRpak0aOQOgqk5JcjTwa8NVF1bVJ/stS5IkaTrNe7dmkrcCJwFXDD9OGq6TJEnSIpt35Ax4NnDwpjk1k5wDfBX44z4LkyRJmkZdn3O208jrB/ZQhyRJkug2cvZW4KtJPsfgcRq/hqNmkiRJvehyQ8AHk1wIHMIgnL2xqn7Qd2GSJEnTqMvIGVV1A7Cq51okSZKmnnNrSpIkNcRwJkmS1JCJ4SzJNkm+sVTFSJIkTbuJ4Wz4bLOvJdl7ieqRJEmaal1uCNgNWJPkX4ENm1ZW1dG9VSVJkjSluoSzv+i9CkmSJAHdnnN2UZKHA/tV1b8kuT+wov/SJEmSpk+Xic//K/BR4H8NV+0BfKzHmiRJkqZWl0dpvAZ4MnALQFV9C3hon0VJkiRNqy7h7Laqun3TQpKVQPVXkiRJ0vTqEs4uSvInwP2SPBP4CPCJfsuSJEmaTl3C2cnAeuDrwO8C5wNv6rMoSZKkaTVvOBs+iPYc4BQGj9U4p6o6ndZMckSSq5KsTXLyhHaHJLkjyQu6Fi5JkrQ16nK35rOBbwOnAe8C1iY5ssN+K4DTgSOBA4BjkxwwR7u3A59aWOmSJElbny4PoT0VOLyq1gIk+Q/APwEXzLPfocDaqrpmuN+5wDHAFbPavRb4B+CQBdQtSZK0VepyzdlNm4LZ0DXATR322wO4bmR53XDdLyXZA3gecOakAyU5IcnqJKvXr1/f4a0lSZK2THOOnCV5/vDlmiTnAx9m8AiN3wIu6XDsjFk3+1q1dwBvrKo7knHNhztVnQWcBTAzM+NjPCRJ0lZr0mnNo0Ze3wg8dfh6PfCgDsdeB+w1srwncP2sNjPAucNgtjPwrCQbq+pjHY4vSZK01ZkznFXV8ffy2JcA+yXZF/g+8CLgxbPeY99Nr5O8D/ikwUySJE2zeW8IGIar1wL7jLavqqMn7VdVG5OcyOAuzBXA2VW1Jskrh9snXmcmSZI0jbrcrfkx4L0MZgW4cyEHr6rzGTy0dnTd2FBWVS9fyLElSZK2Rl3C2b9V1Wm9VyJJkqRO4eydSd4CfBq4bdPKqvpKb1VJkiRNqS7h7LHAccDTuOu0Zg2XJUmStIi6hLPnAY+oqtv7LkaSJGnadZkh4GvATj3XIUmSJLqNnD0M+GaSS7j7NWcTH6UhSZKkhesSzt7SexWSJEkCOoSzqrpoKQqRJElStxkCbuWuCcu3A7YFNlTVjn0WJkmSNI26jJztMLqc5LnAoX0VJEmSNM263K15N8OJyX3GmSRJUg+6nNZ8/sjiNsAMd53mlCRJ0iLqcrfmUSOvNwLfBY7ppRpJkqQp1+Was+OXohBJkiRNCGdJ3jxhv6qqU3qoR5IkaapNGjnbMGbdA4BXAA8BDGeSJEmLbM5wVlWnbnqdZAfgJOB44Fzg1Ln2kyRJ0uabeM1ZkgcDrwNeApwDPL6qfrIUhUmSJE2jSdec/Q/g+cBZwGOr6mdLVpUkSdKUmvQQ2tcDuwNvAq5Pcsvw49YktyxNeZIkSdNl0jVnC549QJIkSfeOAUySJKkhhjNJkqSGGM4kSZIaYjiTJElqiOFMkiSpIYYzSZKkhhjOJEmSGmI4kyRJaojhTJIkqSGGM0mSpIYYziRJkhpiOJMkSWqI4UySJKkhhjNJkqSGGM4kSZIaYjiTJElqiOFMkiSpIYYzSZKkhhjOJEmSGmI4kyRJaojhTJIkqSGGM0mSpIb0Gs6SHJHkqiRrk5w8ZvtLklw+/PhikoP6rEeSJKl1vYWzJCuA04EjgQOAY5McMKvZd4CnVtWBwCnAWX3VI0mStCXoc+TsUGBtVV1TVbcD5wLHjDaoqi9W1U+GixcDe/ZYjyRJUvP6DGd7ANeNLK8brpvLK4ALxm1IckKS1UlWr1+/fhFLlCRJakuf4Sxj1tXYhsnhDMLZG8dtr6qzqmqmqmZ22WWXRSxRkiSpLSt7PPY6YK+R5T2B62c3SnIg8B7gyKr6UY/1SJIkNa/PkbNLgP2S7JtkO+BFwKrRBkn2Bs4Djquqq3usRZIkaYvQ28hZVW1MciLwKWAFcHZVrUnyyuH2M4E3Aw8BzkgCsLGqZvqqSZIkqXWpGnsZWLNmZmZq9erVy12GpC3Y4D+Ds3/3hS3t96Gk9iW5dKEDT84QIEmS1BDDmSRJUkMMZ5IkSQ0xnEmSJDXEcCZJktQQw5kkSVJDDGeSJEkNMZxJkiQ1xHAmSZLUEMOZJElSQwxnkiRJDTGcSZIkNcRwJkmS1BDDmSRJUkMMZ5IkSQ0xnEmSJDXEcCZJktQQw5kkSVJDDGeSJEkNMZxJkiQ1xHAmSZLUEMOZJElSQwxnkiRJDTGcSZIkNcRwJkmS1BDDmSRJUkMMZ5IkSQ0xnEmSJDXEcCZJktQQw5kkSVJDDGeSJEkNMZxJkiQ1xHAmSZLUEMOZJElSQwxnkiRJDTGcSZIkNcRwJkmS1BDDmSRJUkMMZ5IkSQ0xnEmSJDXEcCZJktQQw5kkSVJDDGeSJEkN6TWcJTkiyVVJ1iY5ecz2JDltuP3yJI/vsx5JkqTW9RbOkqwATgeOBA4Ajk1ywKxmRwL7DT9OAP6mr3okSZK2BH2OnB0KrK2qa6rqduBc4JhZbY4B3l8DFwM7Jdmtx5okSZKatrLHY+8BXDeyvA54Yoc2ewA3jDZKcgKDkTWAnyW5anFLXXI7Az9c7iJ0D/ZLe3rsk9xzTe65TmP5b6U99kmbdgYevtCd+gxn437L1Wa0oarOAs5ajKJakGR1Vc0sdx26O/ulPfZJm+yX9tgnbRr2yz4L3a/P05rrgL1GlvcErt+MNpIkSVOjz3B2CbBfkn2TbAe8CFg1q80q4KXDuzafBNxcVTfMPpAkSdK06O20ZlVtTHIi8ClgBXB2Va1J8srh9jOB84FnAWuBnwPH91VPY7aaU7RbGfulPfZJm+yX9tgnbdqsfknVPS7xkiRJ0jJxhgBJkqSGGM4kSZIaYjhbAkkenOT/JPnW8PODJrRdkeSrST65lDVOoy79kmSvJJ9LcmWSNUlOWo5at3ZO9daeDn3ykmFfXJ7ki0kOWo46p818/TLS7pAkdyR5wVLWN4269EmSw5JcNvw7ctF8xzScLY2Tgc9U1X7AZ4bLczkJuHJJqlKXftkIvL6qHg08CXjNmGnIdC841Vt7OvbJd4CnVtWBwCl4QXrvOvbLpnZvZ3BDnnrUpU+S7AScARxdVY8Bfmu+4xrOlsYxwDnD1+cAzx3XKMmewLOB9yxNWVNv3n6pqhuq6ivD17cyCM57LFWBU8Kp3tozb59U1Rer6ifDxYsZPKdS/erybwXgtcA/ADctZXFTqkufvBg4r6quBaiqefvFcLY0Hrbp+W3Dzw+do907gDcAdy5RXdOua78AkGQf4HHAl/svbarMNY3bQtto8Sz0+/0K4IJeKxJ06JckewDPA85cwrqmWZd/K/sDD0pyYZJLk7x0voP2OX3TVEnyL8CuYzb9acf9nwPcVFWXJjlsEUubave2X0aOsz2D/4n+flXdshi16ZcWbao3LZrO3+8khzMIZ0/ptSJBt355B/DGqrrDuWKXRJc+WQk8AXg6cD/gS0kurqqr5zqo4WyRVNUz5tqW5MYku1XVDcNTMeOGNJ8MHJ3kWcB9gR2T/F1V/U5PJU+FRegXkmzLIJh9oKrO66nUaeZUb+3p9P1OciCDyzCOrKofLVFt06xLv8wA5w6D2c7As5JsrKqPLUmF06fr768fVtUGYEOSzwMHAXOGM09rLo1VwMuGr18GfHx2g6r646raczhB6ouAzxrMejdvv2TwG+69wJVV9ddLWNs0caq39szbJ0n2Bs4Djps0AqBFNW+/VNW+VbXP8G/JR4FXG8x61eX318eBX02yMsn9gScyz41/hrOl8TbgmUm+BTxzuEyS3ZOcv6yVTbcu/fJk4DjgacPboC8bjm5qkVTVRmDTVG9XAh/eNNXbpuneGEz1dg2Dqd7eDbx6WYqdEh375M3AQ4Azhv8uVi9TuVOjY79oCXXpk6q6Evhn4HLgX4H3VNU3Jh3X6ZskSZIa4siZJElSQwxnkiRJDTGcSZIkNcRwJkmS1BDDmSRJUkMMZ5IkSQ0xnEkSkOS5Sd6d5ONJfn2565E0vXzOmSSNSPIg4K+q6hXLXYuk6eTImSQBSVYMX74JOH05a5E03Zz4XNLUSvIR4DrgccBnk2wPXFBVX1neyiRNM0fOJE2zxwI/q6rDgZuBZwAvcJ5CScvJa84kTaUk9wWuBXYfTl4sSU1w5EzStHoM8GWDmaTWGM4kTavHApcvdxGSNJvhTNK0MpxJapLXnEmSJDXEkTNJkqSGGM4kSZIaYjiTJElqiOFMkiSpIYYzSZKkhhjOJEmSGmI4kyRJaojhTJIkqSGGM0mSpIb8fwP4re2VbJSdAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "def plotRSquared(r2_values):\n",
    "    # Plot histogram of r^2 per simulation\n",
    "    fig = plt.figure(figsize = (10, 5))\n",
    "    plt.hist(r2_values, color = \"blue\", edgecolor = \"black\", bins = 100)\n",
    "    plt.title(\"Initial $r^{2}$ for 1000 Simulations\")\n",
    "    plt.xlabel(\"$r^{2}$\")\n",
    "    plt.ylabel(\"Number of Simulations\")\n",
    "    \n",
    "    return fig\n",
    "\n",
    "# Plot initial distribution of r^2\n",
    "r2_histogram = plotRSquared(initial_ld_r2)\n",
    "\n",
    "# Save plot\n",
    "r2_histogram.savefig(\"r2_Initial_Distribution.pdf\", bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "91c55ec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the average value across time steps for LD or haplotype frequencies\n",
    "def averageResults(results):\n",
    "    \n",
    "    # Determine the shape of the results array\n",
    "    size = \"2D\" if len(results[0].shape) == 1 else \"3D\"\n",
    "    \n",
    "    if size == \"2D\":\n",
    "        # Get the longest time the simulation lasted\n",
    "        max_time = max([len(x) for x in results])\n",
    "        # Pad the results arrays with None to be the same size\n",
    "        results = [np.pad(results[x], (0, max_time - len(results[x])),\n",
    "                          constant_values = None) for x in range(len(results))]\n",
    "        \n",
    "    else:\n",
    "        # Get the longest time the simulation lasted\n",
    "        max_time = max([x.shape[1] for x in results])\n",
    "        # Pad the results arrays with NaN to be the same size\n",
    "        results = [np.pad(np.array(results[x], dtype = \"object\"),\n",
    "                          [(0, 0), (0, max_time - results[x].shape[1])],\n",
    "                          mode = 'constant', constant_values = float(\"nan\"))\n",
    "                   for x in range(len(results))]\n",
    "\n",
    "    # Record average haplotype frequency, r^2, D' or D\n",
    "    mean_results = np.array([])\n",
    "\n",
    "    # Calculate the average haplotype frequency per time step\n",
    "    for y in range(max_time):\n",
    "        time_step_results = np.array([np.transpose(results[x])[y] for x in range(len(results))])\n",
    "        \n",
    "        if size == \"2D\":\n",
    "            # Remove None or NaN values\n",
    "            time_step_results = [x for x in time_step_results if x != None and not math.isnan(x)]\n",
    "            # Record average\n",
    "            mean_results = np.append(mean_results, np.mean(time_step_results, axis = 0))\n",
    "        \n",
    "        else:\n",
    "            # Record average across 3D arrays, ignoring placeholder NaNs\n",
    "            mean_values = np.nanmean(np.array([np.transpose(results[x])[y]\n",
    "                                            for x in range(len(results))]), axis = 0)\n",
    "            if y == 0:\n",
    "                mean_results = mean_values\n",
    "            else:\n",
    "                mean_results = np.vstack((mean_results, mean_values))\n",
    "        \n",
    "    return mean_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "98cd9fb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def testParameters(alleles, allele_counts, singleton_haploid, \n",
    "                   r = [0.01], s = [0], N = 100, n_simulations = 1000):\n",
    "    \n",
    "    r = np.array(r)\n",
    "    s = np.array(s)\n",
    "    \n",
    "    # Determine whether to test recombination rates or selection coefficients\n",
    "    if len(r) > 1:\n",
    "        test_type = \"recombination\"\n",
    "        rates = r\n",
    "    else:\n",
    "        test_type = \"selection\"\n",
    "        rates = s\n",
    "    \n",
    "    # Create new Wright Fisher simulator\n",
    "    simulator = WrightFisher(N = N, s = s[0])\n",
    "    \n",
    "    # Record results for each recombination rate / selection coefficient for each iteration\n",
    "    all_results = dict(zip(rates, [np.array([]) for i in range(len(rates))]))\n",
    "\n",
    "    # Record haplotype frequencies and LD measures\n",
    "    average_haplotype_freq = {}\n",
    "    average_r2 = {}\n",
    "    average_D = {}\n",
    "    average_D_norm = {}\n",
    "    \n",
    "    r_value = r[0]\n",
    "    s_value = s[0]\n",
    "\n",
    "    # Test each recombination rate / selection coefficient\n",
    "    for i in range(len(rates)):\n",
    "        \n",
    "        if test_type == \"recombination\":\n",
    "            # Set recombination rate\n",
    "            r_value = r[i]\n",
    "            test_value = r_value\n",
    "        else:\n",
    "            # Set selection coefficient\n",
    "            s_value = s[i]\n",
    "            simulator.s = s_value\n",
    "            test_value = s_value\n",
    "\n",
    "        # Run the simulation n times\n",
    "        for j in range(n_simulations):\n",
    "            two_loci_simulation_results = simulator.simulateTwoSites(alleles,\n",
    "                                                                     allele_counts,\n",
    "                                                                     singleton_haploid,\n",
    "                                                                     r = r_value)\n",
    "            # Record the results per simulation\n",
    "            all_results[test_value] = np.append(all_results[test_value],\n",
    "                                                two_loci_simulation_results)\n",
    "\n",
    "        # Get the haplotype frequencies for the recombination rate\n",
    "        average_haplotype_freq[test_value] = averageResults([res[\"haplotype_frequencies\"] for res in all_results[test_value]])\n",
    "        average_r2[test_value] = averageResults([res[\"r2\"] for res in all_results[test_value]][:-1])\n",
    "        average_D[test_value] = averageResults([res[\"D\"] for res in all_results[test_value]][:-1])\n",
    "        average_D_norm[test_value] = averageResults([res[\"D'\"] for res in all_results[test_value]][:-1])\n",
    "        \n",
    "    return {\"results_per_simulation\": all_results,\n",
    "            \"average_haplotype_freq\": average_haplotype_freq,\n",
    "            \"average_r2\": average_r2,\n",
    "            \"average_D\": average_D,\n",
    "            \"average_D_norm\": average_D_norm}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d277fa47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recombination rates and selection coefficients to test\n",
    "test_r = [0.01, 0.02, 0.05]\n",
    "test_s = [0.02, 0.005, -0.005, -0.02]\n",
    "    \n",
    "# # Test different recombination rates\n",
    "# recombination_results = testParameters(initial_alleles,\n",
    "#                                        initial_allele_counts,\n",
    "#                                        singleton,\n",
    "#                                        r = test_r,\n",
    "#                                        s = [0],\n",
    "#                                        n_simulations = 1000)\n",
    "\n",
    "# Test different selection coefficients\n",
    "selection_results = testParameters(initial_alleles,\n",
    "                                   initial_allele_counts,\n",
    "                                   singleton,\n",
    "                                   r = [0.01],\n",
    "                                   s = test_s,\n",
    "                                   n_simulations = 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abb9bb2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot D, D' and r^2 over all simulations and the average\n",
    "def plotLinkageDisequilibrium(colours, linestyles, all_results, r = [0.01], s = [0]):\n",
    "    \n",
    "    r = np.array(r)\n",
    "    s = np.array(s)\n",
    "    \n",
    "    # Determine whether to plot recombination rates or selection coefficients\n",
    "    if len(r) > 1:\n",
    "        test_type = \"recombination\"\n",
    "        rates = r\n",
    "    else:\n",
    "        test_type = \"selection\"\n",
    "        rates = s\n",
    "        \n",
    "    # Determine the number of simulations\n",
    "    n_simulations = len(all_results[\"results_per_simulation\"][rates[0]])\n",
    "\n",
    "    # Create subplots of LD over time for the recombination rates\n",
    "    ld_fig, axes = plt.subplots(3, 2, figsize = (15, 10), sharex = True)\n",
    "    ld_fig.suptitle(\"Linkage Disequilibrium over Time for \" + str(n_simulations) + \" Simulations\",\n",
    "                      fontsize = 13, y = 0.95)\n",
    "\n",
    "    # Iterate through recombination rates / selection coefficients\n",
    "    for value in rates:\n",
    "        # Get LD results\n",
    "        results_per_simulation = all_results[\"results_per_simulation\"][value]\n",
    "        average_r2 = all_results[\"average_r2\"][value]\n",
    "        average_D = all_results[\"average_D\"][value]\n",
    "        average_D_norm = all_results[\"average_D_norm\"][value]\n",
    "\n",
    "        # Overlay D, D' and r^2 results for each simulation by adding transparent lines\n",
    "        for n in range(n_simulations):\n",
    "            axes[0,0].plot(results_per_simulation[n][\"D\"],\n",
    "                           color = colours[value],\n",
    "                           linestyle = linestyles[value],\n",
    "                           alpha = 0.3)\n",
    "            axes[1,0].plot(results_per_simulation[n][\"D'\"],\n",
    "                           color = colours[value],\n",
    "                           linestyle = linestyles[value],\n",
    "                           alpha = 0.3)\n",
    "            axes[2,0].plot(results_per_simulation[n][\"r2\"],\n",
    "                           color = colours[value],\n",
    "                           linestyle = linestyles[value],\n",
    "                           alpha = 0.3)\n",
    "\n",
    "        # Plot average D, D' and r^2 for recombination rate / selection coefficient\n",
    "        axes[0,1].plot(average_D, color = colours[value],\n",
    "                       linestyle = linestyles[value], label = value)\n",
    "        axes[1,1].plot(average_D_norm, color = colours[value],\n",
    "                       linestyle = linestyles[value])\n",
    "        axes[2,1].plot(average_r2, color = colours[value],\n",
    "                       linestyle = linestyles[value])\n",
    "\n",
    "    # Set axis labels and titles\n",
    "    axes[0,0].set_ylabel(\"Coefficient of LD (D)\")\n",
    "    axes[1,0].set_ylabel(\"Normalised coefficient of LD (D')\")\n",
    "    axes[2,0].set_ylabel(\"Genetic Correlation ($r^{2}$)\")\n",
    "    axes[2,0].set_xlabel(\"Time (t)\")\n",
    "    axes[2,1].set_xlabel(\"Time (t)\")\n",
    "    axes[0,0].set_title(\"All Results\")\n",
    "    axes[0,1].set_title(\"Average\")\n",
    "\n",
    "    # Position legend and show plot\n",
    "    ld_fig.legend(loc = \"upper right\", bbox_to_anchor = (1, 0.88),\n",
    "                  title = \"Recombination\\nRate (r)\\n\" if test_type == \"recombination\"\n",
    "                  else \"Selection\\nCoefficient (s)\\n\")\n",
    "    plt.show()\n",
    "    \n",
    "    return ld_fig\n",
    "\n",
    "# Set plot styles for values to test\n",
    "r_colours = {0.01: \"yellowgreen\", 0.02: \"aqua\", 0.05: \"indigo\"}\n",
    "s_colours = {0.02: \"lightblue\", 0.005: \"orange\", -0.005: \"fuchsia\", -0.02: \"navy\"}\n",
    "r_linestyles = {0.05: \"solid\", 0.01: \"solid\", 0.02: \"solid\"}\n",
    "s_linestyles = {0.02: \"solid\", 0.005: \"solid\", -0.005: \"solid\", -0.02: \"solid\"}\n",
    "\n",
    "# Plot LD measures\n",
    "r_ld_fig = plotLinkageDisequilibrium(r_colours, r_linestyles, recombination_results, r = test_r)\n",
    "s_ld_fig = plotLinkageDisequilibrium(s_colours, s_linestyles, selection_results, s = test_s)\n",
    "\n",
    "r_ld_fig.savefig(\"LD_Over_Time_Recombination_Rates.pdf\", bbox_inches = \"tight\")\n",
    "s_ld_fig.savefig(\"LD_Over_Time_Selection_Coefficients.pdf\", bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f062460b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotHaplotypeFrequency(colours, all_results, haplotype_names, r = [0.01], s = [0]):\n",
    "    \n",
    "    r = np.array(r)\n",
    "    s = np.array(s)\n",
    "    \n",
    "    # Determine whether to plot recombination rates or selection coefficients\n",
    "    if len(r) > 1:\n",
    "        test_type = \"recombination\"\n",
    "        rates = r\n",
    "    else:\n",
    "        test_type = \"selection\"\n",
    "        rates = s\n",
    "    \n",
    "    # Determine the number of simulations\n",
    "    n_simulations = len(all_results[\"results_per_simulation\"][rates[0]])\n",
    "    \n",
    "    # Plot haplotype frequency for the recombination rates\n",
    "    haplotype_fig, axes = plt.subplots(len(rates), 2, figsize = (15, 10), sharex = True)\n",
    "    haplotype_fig.suptitle(\"Haplotype Frequency over Time for \" + str(n_simulations) + \" Simulations\",\n",
    "                 fontsize = 13, y = 0.95)\n",
    "\n",
    "    # Add plot for each recombination rate\n",
    "    for i in range(len(rates)):\n",
    "        value = rates[i]\n",
    "        \n",
    "        results_per_simulation = all_results[\"results_per_simulation\"][value]\n",
    "        average_haplotype_freq = all_results[\"average_haplotype_freq\"][value]\n",
    "\n",
    "        # Overlay results for each simulation by adding transparent lines\n",
    "        for n in range(n_simulations):\n",
    "            for j in range(len(haplotype_names)):\n",
    "                haplotype = haplotype_names[j]\n",
    "                axes[i,0].plot(results_per_simulation[n][\"haplotype_frequencies\"][j],\n",
    "                               color = colours[haplotype], alpha = 0.3)\n",
    "\n",
    "        # Plot average per haplotype\n",
    "        for k in range(len(haplotype_names)):\n",
    "            haplotype = haplotype_names[k]\n",
    "            haplotype_freq = np.transpose(average_haplotype_freq)[k]\n",
    "            axes[i,1].plot(haplotype_freq, color = colours[haplotype],\n",
    "                           label = haplotype if i == 1 else \"\")\n",
    "            \n",
    "        # Set y-axis labels\n",
    "        axes[i,0].set_ylabel(test_type[0] + \" = \" + str(rates[i]))\n",
    "\n",
    "    # Set x-axis labels and titles\n",
    "    axes[2,0].set_xlabel(\"Time (t)\")\n",
    "    axes[2,1].set_xlabel(\"Time (t)\")\n",
    "    axes[0,0].set_title(\"All Results\")\n",
    "    axes[0,1].set_title(\"Average\")\n",
    "\n",
    "    # Position legend and show plot\n",
    "    haplotype_fig.legend(loc = \"upper right\", bbox_to_anchor = (0.98, 0.88), title = \"Haplotype\")\n",
    "    plt.show()\n",
    "    \n",
    "    return haplotype_fig\n",
    "\n",
    "# Set plot colours\n",
    "r_haplotype_colours = {\"AB\": \"blue\", \"Ab\": \"orange\", \"aB\": \"green\", \"ab\": \"red\"}\n",
    "s_haplotype_colours = {\"AB\": \"orangered\", \"Ab\": \"orchid\", \"aB\": \"gold\", \"ab\": \"chartreuse\"}\n",
    "\n",
    "# Plot haplotype frequency over time for different r and s values\n",
    "r_haplotype_fig = plotHaplotypeFrequency(r_haplotype_colours, recombination_results, all_haplotypes, test_r)\n",
    "s_haplotype_fig = plotHaplotypeFrequency(s_haplotype_colours, selection_results, all_haplotypes, test_s)\n",
    "\n",
    "# Save the plots\n",
    "r_haplotype_fig.savefig(\"Haplotypes_Over_Time_Recombination_Rates.pdf\", bbox_inches = \"tight\")\n",
    "s_haplotype_fig.savefig(\"Haplotypes_Over_Time_Selection_Coefficients.pdf\", bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8640237",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Record the haplotypes and allele frequencies for the last generation in each simulation\n",
    "final_haplotypes = dict(zip(test_s, [np.array([]) for i in range(len(test_s))]))\n",
    "final_allele_counts = dict(zip(test_s, [{} for i in range(len(test_s))]))\n",
    "\n",
    "# Iterate over s values\n",
    "for i in range(len(test_s)):\n",
    "    s_value = test_s[i]\n",
    "    \n",
    "    # Get frequencies per simulation\n",
    "    for n in range(len(selection_results[\"results_per_simulation\"][s_value])):\n",
    "        # Find the haplotype frequencies at the final generation\n",
    "        final_haplotype_freq = selection_results[\"results_per_simulation\"][s_value][n][\"haplotype_frequencies\"][:, -1]\n",
    "        # Record the final frequencies\n",
    "        if n == 0:\n",
    "            final_haplotypes[s_value] = final_haplotype_freq\n",
    "        else:\n",
    "            final_haplotypes[s_value] = np.vstack((final_haplotypes[s_value],final_haplotype_freq))\n",
    "            \n",
    "    # Calculate allele counts\n",
    "    final_allele_counts[s_value][\"A\"] = final_haplotypes[s_value][:, 0] + final_haplotypes[s_value][:, 1]\n",
    "    final_allele_counts[s_value][\"a\"] = final_haplotypes[s_value][:, 2] + final_haplotypes[s_value][:, 3]\n",
    "    final_allele_counts[s_value][\"B\"] = final_haplotypes[s_value][:, 0] + final_haplotypes[s_value][:, 2]\n",
    "    final_allele_counts[s_value][\"b\"] = final_haplotypes[s_value][:, 1] + final_haplotypes[s_value][:, 3]\n",
    "    \n",
    "    # Calculate mean of each allele\n",
    "    allele_means = [np.mean(a) for a in list(final_allele_counts[s_value].values())]\n",
    "    \n",
    "    if i == 0:\n",
    "        allele_mean_data = allele_means\n",
    "    else:\n",
    "        allele_mean_data = np.vstack((allele_mean_data, [np.mean(a) for a in list(final_allele_counts[s_value].values())]))\n",
    "        \n",
    "# Create dataframe for plotting\n",
    "allele_mean_df = pd.DataFrame(np.transpose(np.vstack((list(final_allele_counts[s_value].keys()),\n",
    "                                                      allele_mean_data))),\n",
    "                              columns = [\"Allele\"] + test_s)\n",
    "# Convert mean allele columns to numeric\n",
    "allele_mean_df[test_s] = allele_mean_df[test_s].apply(pd.to_numeric)\n",
    "\n",
    "# View the data\n",
    "print(allele_mean_df)\n",
    "\n",
    "# Plot grouped bar chart\n",
    "allele_mean_fig = allele_mean_df.plot(x = \"Allele\", kind = \"bar\", stacked = False,\n",
    "                                      logy = True, title = \"Mean Allele Frequencies\",\n",
    "                                      color = list(s_colours.values()),\n",
    "                                      figsize = (10, 5))\n",
    "allele_mean_fig.legend(loc = \"upper right\", bbox_to_anchor = (1.33, 0.9),\n",
    "                       title = \"Selection\\nCoefficient (s)\")\n",
    "plt.show()\n",
    "\n",
    "# Save the plot\n",
    "allele_mean_fig.get_figure().savefig(\"Mean_Allele_Frequencies_Selection_Coefficients.pdf\",\n",
    "                                     bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8812f32",
   "metadata": {},
   "source": [
    "Extra plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4baaf5b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(4, 1, figsize = (15, 10), sharex = True)\n",
    "fig.suptitle(\"Allele Frequency over Time for \" + str(len(final_allele_counts[s_value][all_alleles[0]])) + \" Simulations\",\n",
    "             fontsize = 13, y = 0.95)\n",
    "\n",
    "for i in range(len(final_haplotypes.keys())):\n",
    "    s = list(final_allele_counts.keys())[i]\n",
    "    for j in range(4):\n",
    "        allele = all_alleles[j]\n",
    "        axes[i].hist(final_allele_counts[s][allele], label = allele if i == 0 else \"\",\n",
    "                             bins = 100, alpha = 0.5)\n",
    "        \n",
    "    # Set y-axis\n",
    "    axes[i].set_ylabel(\"s = \" + str(s))\n",
    "    axes[i].set_yscale(\"log\")\n",
    "\n",
    "# Position legend and show plot\n",
    "fig.legend(loc = \"upper right\", bbox_to_anchor = (0.98, 0.88), title = \"Haplotype\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "574adeb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(4, 1, figsize = (15, 10), sharex = True)\n",
    "fig.suptitle(\"Haplotype Frequency over Time for Simulations\",\n",
    "             fontsize = 13, y = 0.95)\n",
    "\n",
    "for i in range(len(final_haplotypes.keys())):\n",
    "    s = list(final_haplotypes.keys())[i]\n",
    "    for j in range(4):\n",
    "        haplotype = all_haplotypes[j]\n",
    "        axes[i].hist(final_haplotypes[s][:, j], label = haplotype if i == 0 else \"\",\n",
    "                             bins = 100, alpha = 0.5)\n",
    "        \n",
    "    # Set y-axis\n",
    "    axes[i].set_ylabel(\"s = \" + str(s))\n",
    "    axes[i].set_yscale(\"log\")\n",
    "\n",
    "# Position legend and show plot\n",
    "fig.legend(loc = \"upper right\", bbox_to_anchor = (0.98, 0.88), title = \"Haplotype\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
